File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/NetWork/CmdDispatch/NetModuleDispatchHandle.cpp
#include "MyProject.h"
#include "NetModuleDispatchHandle.h"
#include "MByteBuffer.h"
#include "UtilContainers.h"
#include "Prequisites.h"
#include "UtilStr.h"
#include "NetCmdDispatchHandle.h"
#include "AddOnceEventDispatch.h"
#include "CmdDispatchInfo.h"
#include "MClassInfo.h"
#include "MClassMacros.h"
#include "MClassFactory.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(NetModuleDispatchHandle, GObject)

NetModuleDispatchHandle::NetModuleDispatchHandle()
{
	
}

NetModuleDispatchHandle::~NetModuleDispatchHandle()
{
	
}

void NetModuleDispatchHandle::init()
{

}

void NetModuleDispatchHandle::dispose()
{

}

void NetModuleDispatchHandle::addCmdHandle(int cmdId, EventDispatchDelegate handle)
{
	if (!this->mId2DispatchDic.containsKey(cmdId))
	{
		this->mId2DispatchDic[cmdId] = MY_NEW AddOnceEventDispatch();
	}

	this->mId2DispatchDic[cmdId]->addEventHandle(handle);
}

void NetModuleDispatchHandle::removeCmdHandle(int cmdId, EventDispatchDelegate handle)
{
	if (this->mId2DispatchDic.containsKey(cmdId))
	{
		this->mId2DispatchDic[cmdId]->removeEventHandle(handle);
	}
	else
	{
		// 日志
	}
}

void NetModuleDispatchHandle::handleMsg(CmdDispatchInfo* cmdDispInfo)
{
	if (this->mId2DispatchDic.containsKey(cmdDispInfo->byCmd))
    {
		GLogSys->log(UtilStr::Format("处理消息: byCmd = {0},  byParam = {1}", cmdDispInfo->byCmd, cmdDispInfo->byParam));
		this->mId2DispatchDic[cmdDispInfo->byCmd]->dispatchEvent(cmdDispInfo);
    }
    else
    {
		GLogSys->log(UtilStr::Format("消息没有处理: byCmd = {0},  byParam = {1}", cmdDispInfo->byCmd, cmdDispInfo->byParam));
    }

    //if(mLuaCSBridgeNetDispHandle != null)
    //{
    //    mLuaCSBridgeNetDispHandle.handleMsg(msg, byCmd, byParam);
    //}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Network/CmdDispatch/NetModuleDispatchHandle.lua
local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "NetModuleDispatchHandle";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mId2DispatchDic = GlobalNS.new(GlobalNS.MDictionary);
    --self.mCmdDispInfo = GlobalNS.new(GlobalNS.CmdDispatchInfo);
end

function M:addCmdHandle(cmdId, pThis, func, eventId)
    if (not self.mId2DispatchDic:containsKey(cmdId)) then
        local disp = GlobalNS.new(GlobalNS.AddOnceEventDispatch);
        self.mId2DispatchDic:add(cmdId, disp);
    end

    self.mId2DispatchDic:value(cmdId):addEventHandle(pThis, func, eventId);
end

function M:removeCmdHandle(cmdId, pThis, func, eventId)
    if(not self.mId2DispatchDic.ContainsKey(cmdId)) then
        GCtx.mLogSys:log("Cmd Handle Not Register");
    end

    self.mId2DispatchDic[cmdId]:removeEventHandle(pThis, func, eventId);
end

function M:handleMsg(msg)
    --local byCmd = msg.mByCmd;
    --byCmd = msg.readUnsignedInt8(byCmd);
    --local byParam = msg.mByParam;
    --byParam = msg.readUnsignedInt8(byParam);
    --msg.setPos(0);
    GCtx.mLogSys:log("NetModuleDispatchHandle Start handleMsg", GlobalNS.LogTypeId.eLogCommon);
	
    if(self.mId2DispatchDic:containsKey(msg.byCmd)) then
        GCtx.mLogSys:log("NetModuleDispatchHandle In handleMsg", GlobalNS.LogTypeId.eLogCommon);
        --self.mCmdDispInfo.bu = msg.mMsgBody;
        --self.mCmdDispInfo.byCmd = byCmd;
        --self.mCmdDispInfo.byParam = byParam;
        self.mId2DispatchDic:value(msg.byCmd):dispatchEvent(msg);
    else
        
    end
end

return M;